File: 1530-number-of-good-leaf-nodes-pairs/1530-number-of-good-leaf-nodes-pairs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countPairs(TreeNode* root, int distance) {
        traverseTree(root, graph, leafs);
        int ans = 0;

        for (auto node : leafs)
        {
            unordered_set<TreeNode*> passed;
            passed.insert(node);
            queue<TreeNode*> node_queue;
            node_queue.push(node);
    
            // using the idea from editorial, do the bfs in a for-loop instead of a while-loop
            for (int dist=0; dist<distance+1; dist++)
            {
                int n = node_queue.size();
                for (int i=0; i<n; i++)
                {
                    TreeNode* curr = node_queue.front();
                    node_queue.pop();
                    passed.insert(curr);

                    if (leafs.contains(curr) && curr!=node)
                        ans += 1;
                    else
                    {
                        for (auto next : graph[curr])
                        {
                            if (!passed.contains(next))
                                node_queue.push(next);
                        }
                    }
                }
            }
        }

        return ans/2;
    }

private:
    unordered_map<TreeNode*, vector<TreeNode*>> graph;
    unordered_set<TreeNode*> leafs;

    void traverseTree(TreeNode* node, unordered_map<TreeNode*, vector<TreeNode*>>& graph, unordered_set<TreeNode*>& leafs)
    {
        if (!node->left && !node->right) 
        {
            leafs.insert(node);
            return;
        }

        if (node->left)
        {
            graph[node->left].push_back(node);
            graph[node].push_back(node->left);
            traverseTree(node->left, graph, leafs);
        }

        if (node->right)
        {
            graph[node->right].push_back(node);
            graph[node].push_back(node->right);
            traverseTree(node->right, graph, leafs);
        }

        return;
    }
};

File: 1530-number-of-good-leaf-nodes-pairs/1530-number-of-good-leaf-nodes-pairs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        # store the Tree into a graph
        graph = {}
        graph[root] = []
        leaf = set()

        # traverse the tree, find the neighbors and find the leaf nodes
        def traverseTree(root):
            # if a leaf node is reached
            if not root.left and not root.right:
                leaf.add(root)
                return

            if root.left:
                if root.left not in graph:
                    graph[root.left] = []
                # add the neighbor into the graph
                graph[root.left].append(root)
                graph[root].append(root.left)
                traverseTree(root.left)

            if root.right:
                if root.right not in graph:
                    graph[root.right] = []
                graph[root.right].append(root)
                graph[root].append(root.right)
                traverseTree(root.right)
        
        traverseTree(root)
        ans = 0

        # go through all the leaf nodes in bfs
        for x in leaf:
            # keep a record of which node is passed
            passed = set()
            queue = deque()
            queue.append(x)
            passed.add(x)
            dist = 0

            # bfs
            while len(queue)!=0:
                n = len(queue)
                for y in range(n):
                    node = queue.popleft()
                    passed.add(node)
                    if node in leaf and dist!=0:
                        print(node.val, x.val)
                        ans += 1
                    else:
                        for z in graph[node]:
                            if z not in passed:
                                queue.append(z)
                if distance==dist:
                    break
                dist+=1
        return ans//2

